cdap-app-fabric/src/main/java/io/cdap/cdap/support/services/SupportBundleService.java

```diff
@@ -0,0 +1,223 @@
+    File basePath = new File(localDir, uuid);
+    SupportBundleJob supportBundleJob =
+      new SupportBundleJob(
+        supportBundleTaskFactoryList,
+        executorService,
+        cConf,
+        supportBundleStatus);
+    String namespaceId = supportBundleConfiguration.getNamespaceId();
+    SupportBundleState supportBundleState = new SupportBundleState(supportBundleConfiguration);
+    if (namespaceId != null) {
+      NamespaceId namespace = new NamespaceId(namespaceId);
+      if (!namespaceQueryClient.exists(namespace)) {
+        throw new NamespaceNotFoundException(namespace);
+      }
+    }
+    try {
+      Long startTs = System.currentTimeMillis();
+      supportBundleStatus.setBundleId(uuid);
+      supportBundleStatus.setStartTimestamp(startTs);
+
+      supportBundleStatus.setParameters(supportBundleConfiguration);
+
+      List<String> namespaceList = new ArrayList<>();
+      if (namespaceId == null) {
+        namespaceList.addAll(
+          namespaceQueryClient.list().stream()
+            .map(NamespaceMeta::getName)
+            .collect(Collectors.toList()));
+      } else {
+        namespaceList.add(namespaceId);
+      }
+      // Puts all the files under the uuid path
+      File baseDirectory = new File(localDir);
+      DirUtils.mkdirs(baseDirectory);
+      int fileCount = DirUtils.list(baseDirectory).size();
+
+      // We want to keep consistent number of bundle to provide to customer
+      if (fileCount >= folderMaxNumber) {
+        File oldFilesDirectory = getOldestFolder(baseDirectory);
+        deleteOldFolders(oldFilesDirectory);
+      }
+      DirUtils.mkdirs(basePath);
+      supportBundleState.setUuid(uuid);
+      supportBundleState.setBasePath(basePath.getPath());
+      supportBundleState.setNamespaceList(namespaceList);
+      supportBundleState.setSupportBundleJob(supportBundleJob);
+      executorService.execute(() -> supportBundleJob.generateBundle(supportBundleState));
+    } catch (Exception e) {
+      LOG.error("Can not generate support bundle ", e);
+      supportBundleStatus.setStatus(CollectionState.FAILED);
+      supportBundleStatus.setStatusDetails(e.getMessage());
+      supportBundleStatus.setFinishTimestamp(System.currentTimeMillis());
+      addToStatus(supportBundleStatus, basePath.getPath());
+      throw new Exception("Can not generate support bundle ", e);
+    }
+    return uuid;
+  }
+
+  /**
+   * Gets oldest folder from the root directory
+   */
+  private File getOldestFolder(File baseDirectory) throws RuntimeException {
+    File[] uuidFiles = baseDirectory.listFiles(
+      (dir, name) -> !name.startsWith(".") && !dir.isHidden() && dir.isDirectory());
+    return Collections.min(Arrays.asList(uuidFiles),
+                           Comparator.<File, Boolean>comparing(
+                               f1 -> {
+                                 try {
+                                   return getSingleBundleJson(f1).getStatus()
```

Comment:
Why using boolean as the comparison key? Is it because you only want non failed file? Shouldn't it be a filter instead?

Reply:
no we want to find the failed file first to remove it from the directory, if no failed file then we compare with lastmodified file list